test/extensions/filters/udp/udp_proxy/udp_proxy_integration_test.cc

```diff
@@ -0,0 +1,182 @@
+    // modification black to the abstraction layer so it will work for multiple platforms.
+    RELEASE_ASSERT(fcntl(socket_->ioHandle().fd(), F_SETFL, 0) != -1, "");
+  }
+
+  void write(const std::string& buffer, const Network::Address::Instance& peer) {
+    const auto rc = Network::Utility::writeToSocket(socket_->ioHandle(), Buffer::OwnedImpl(buffer),
+                                                    nullptr, peer);
+    ASSERT_EQ(rc.rc_, buffer.length());
+  }
+
+  void recv(Network::UdpRecvData& datagram) {
+    SyncPacketProcessor processor(datagram);
+    const auto rc =
+        Network::Utility::readFromSocket(socket_->ioHandle(), *socket_->localAddress(), processor,
+                                         time_system_.monotonicTime(), nullptr);
+    ASSERT_TRUE(rc.ok());
+  }
+
+private:
+  struct SyncPacketProcessor : public Network::UdpPacketProcessor {
+    SyncPacketProcessor(Network::UdpRecvData& data) : data_(data) {}
+
+    void processPacket(Network::Address::InstanceConstSharedPtr local_address,
+                       Network::Address::InstanceConstSharedPtr peer_address,
+                       Buffer::InstancePtr buffer, MonotonicTime receive_time) override {
+      data_.addresses_.local_ = std::move(local_address);
+      data_.addresses_.peer_ = std::move(peer_address);
+      data_.buffer_ = std::move(buffer);
+      data_.receive_time_ = receive_time;
+    }
+    uint64_t maxPacketSize() const override { return Network::MAX_UDP_PACKET_SIZE; }
+
+    Network::UdpRecvData& data_;
+  };
+
+  Event::TestTimeSystem& time_system_;
+  const Network::SocketPtr socket_;
+};
+
+class UdpProxyIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                public BaseIntegrationTest {
+public:
+  UdpProxyIntegrationTest() : BaseIntegrationTest(GetParam(), configToUse()) {}
+
+  static std::string configToUse() {
+    return ConfigHelper::BASE_UDP_LISTENER_CONFIG + R"EOF(
+    listener_filters:
+      name: envoy.filters.udp_listener.udp_proxy
+      typed_config:
+        '@type': type.googleapis.com/envoy.config.filter.udp.udp_proxy.v2alpha.UdpProxyConfig
+        cluster: cluster_0
+      )EOF";
+  }
+
+  /**
+   * Initializer for an individual test.
+   */
+  void SetUp() override {
+    udp_fake_upstream_ = true;
+    BaseIntegrationTest::initialize();
+  }
+
+  /**
+   *  Destructor for an individual test.
+   */
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  void requestResponseWithListenerAddress(const Network::Address::Instance& listener_address) {
+    // Send datagram to be proxied.
+    UdpSyncClient client(timeSystem(), version_);
+    client.write("hello", listener_address);
+
+    // Wait for the upstream datagram.
+    Network::UdpRecvData request_datagram;
+    ASSERT_TRUE(fake_upstreams_[0]->waitForUdpDatagram(request_datagram));
+    EXPECT_EQ("hello", request_datagram.buffer_->toString());
```

Comment:
Can you also verify the source and dest addresses of this packet? 
Ditto to the response packet.

Reply:
Sorry, I misunderstood that the proxying envoy was transparent to upstream host. But actually the incoming packets' source and destination addresses are different from what client sent.

Reply:
Yeah in this version it's not transparent. I think we could make it transparent later by using raw sockets. This is why I was asking you the question about writing source port, etc.